feature/healthy_meals

tasneem-hakeem opened the pull request:
Implementing the get quick and healthy recipes use case and the structure of ui

Commit:
implement GetHealthyRecipesUseCase

Commit:
implement ui structure

Commit:
add recipe formatter class for more readable output

Commit:
add null safety checks

Commit:
add use case module

Commit:
add low average ratio

Commit:
Merge remote-tracking branch 'origin/develop' into feature/healthy_meals

# Conflicts:
#	src/main/kotlin/Main.kt
#	src/main/kotlin/di/dataModule.kt
#	src/main/kotlin/di/useCaseModule.kt
#	src/main/kotlin/ui/FoodChangeMoodUi.kt

Commit:
resolve conflicts

Commit:
update healthy recipe use case class to sort by the average of the nutrition values

Commit:
Merge remote-tracking branch 'origin/develop' into feature/healthy_meals

# Conflicts:
#	src/main/kotlin/di/dataModule.kt
#	src/main/kotlin/di/useCaseModule.kt
#	src/main/kotlin/ui/FoodChangeMoodUi.kt

tasneem-hakeem committed:
Merge branch 'develop' into feature/healthy_meals

Commit:
solve koin modules problems

Commit:
Merge remote-tracking branch 'origin/feature/healthy_meals' into feature/healthy_meals

# Conflicts:
#	src/main/kotlin/ui/FoodChangeMoodUi.kt

Commit:
add missed comma

Commit:
add null safety checks

Commit:
Merge remote-tracking branch 'origin/develop' into feature/healthy_meals

# Conflicts:
#	src/main/kotlin/Main.kt
#	src/main/kotlin/di/dataModule.kt
#	src/main/kotlin/di/useCaseModule.kt
#	src/main/kotlin/ui/FoodChangeMoodUi.kt
#	src/main/kotlin/ui/uiModule.kt

Commit:
resolve merge conflicts

iZyadNasser commented on src/main/kotlin/logic/HealthyRecipesUseCase.kt:
```diff
@@ -0,0 +1,40 @@
+        val saturatedFat = recipe.nutrition?.saturatedFat ?: ZERO
+        val carbs = recipe.nutrition?.carbohydrates ?: ZERO
+
+        return (totalFat + saturatedFat + carbs) / NUMBER_OF_ELEMENTS
```
You didn't take the difference in the unit significance for each one of the three measures into consideration (for example: the average carbs that are recommended daily are about 130g, saturated fat is 30g, total fat about 50g, so not all of them hold the same strength in calculation),

A way to solve this is to multiply each of them by the ratio of their average value over the sum of all averages, so sumOfAvgs = 130 + 30 +50 = 210, so totalFatRatio = 50/210, saturatedFatRatio = 30/210, and carbs = 130/210, so multiply every value with its ratio before summing,

also, you don't need to divide by the number of elements (3) as it is constant in all cases anyway so it won't affect the comparison for sorting

iZyadNasser commented on src/main/kotlin/ui/FoodChangeMoodUi.kt:
```diff
@@ -79,6 +83,14 @@ class FoodChangeMoodUi(
     }
+    private fun launchHealthyRecipes() {
+        println("Enter the number of meals you want")
+        getUserInput()?.let { count ->
```
The user could enter a minus number which would result in a crash, handle this case (I prefer you handle all this validation in the validator class but if find that too much then no problem you can handle it here)

Commit:
fix healthy recipe use case
- sort recipes compared by saturated fat, total fat and carbohydrates
- add validation for recipe count input in ui

Commit:
Merge remote-tracking branch 'origin/develop' into feature/healthy_meals

# Conflicts:
#	src/main/kotlin/di/dataModule.kt
#	src/main/kotlin/ui/FoodChangeMoodUi.kt

Commit:
resolve merge conflicts

iZyadNasser reviewed (commented):
continue modifying the comments

Commit:
update the logic of sorting to sort by one value: the sum of total fat, saturated fat and carbohydrates

Commit:
Merge remote-tracking branch 'origin/develop' into feature/healthy_meals

# Conflicts:
#	src/main/kotlin/di/dataModule.kt
#	src/main/kotlin/ui/FoodChangeMoodUi.kt

Commit:
resolve merge conflicts
